src/signature.c

```diff
@@ -24,198 +24,386 @@
-	if (!result) {
-		unlink(sig_filename);
+	data_size = st.st_size;
+	data = malloc(data_size);
+	count = fread(data, 1, data_size, fp_data);
+	if (count != data_size) {
+		fprintf(stderr, "Failed to read full data file\n");
+		goto error;
 	}
-	free(sig_filename);
-	free(sig_url);
-	return result || (IMPL == IMPL_FORGIVE);
+
+	/* Load the signature file which is in PKCS7 format */
+	sigbio = BIO_new_mem_buf(signature, sig_len);
+	if (!sigbio) {
+		fprintf(stderr, "Failed to read signature into buffer\n");
+		goto error;
+	}
+
+	/* Read the PKCS7 file which is in DER format, hence the d2i function */
+	p7 = d2i_PKCS7_bio(sigbio, NULL);
+	if (p7 == NULL) {
+		fprintf(stderr, "NULL PKCS7 File\n");
+		goto error;
+	}
+
+	/* Load the data to be verified */
+	databio = BIO_new_mem_buf(data, data_size);
+	if (!databio) {
+		fprintf(stderr, "Failed to read data file into buffer\n");
+		goto error;
+	}
+
+	/* Initialize the data that will be passed into verify with the PKCS7 */
+	indata = PKCS7_dataInit(p7, databio);
+	if (!indata) {
+		fprintf(stderr, "Failed to initialize data buffer for verification\n");
+		goto error;
+	}
+
+	/* Push our trust cert(s) to the stack, which is a set of certificates
+	 * in which to search for the signer's cert. */
+	x509_stack = sk_X509_new_null();
+	if (!x509_stack) {
+		fprintf(stderr, "Failed to initialize X509 certificate stack\n");
+		goto error;
+	}
+	sk_X509_push(x509_stack, cert);
+
+	/* Verify the signature, outdata can be NULL because we don't use it */
+	ret = PKCS7_verify(p7, x509_stack, store, indata, NULL, 0);
+	if (ret == 1) {
+		printf("VERIFY SUCCESS!\n");
+		return true;
+	}
+error:
+	if (signature) {
+		free(signature);
+	}
+	if (data) {
+		free(data);
+	}
+	ERR_print_errors_fp(stderr);
+	fprintf(stderr, "Signature check failed\n");
+	return false;
 }
 
-void signature_delete(const char *data_filename)
+/* This function makes sure the certificate is still valid by not having any
+ * compromised certificates in the chain.
+ * If there is no Certificate Revocation List (CRL) it may be that the private
+ * keys have not been compromised or the CRL has not been generated by the
+ * Certificate Authority (CA)
+ *
+ * returns: true if certificate is valid, false otherwise */
+static bool validate_certificate(void)
 {
-	char *sig_filename;
+	FILE *fp = NULL;
+	X509_LOOKUP *lookup = NULL;
+	X509_STORE_CTX *verify_ctx = NULL;
+
+	/* CRL and Chains are not required for the current setup, but we may
+	 * implement them in the future */
+	if (!crl) {
+		printf("No certificate revocation list provided\n");
+	}
+	if (!chain) {
+		printf("No certificate chain provided\n");
+	}
 
-	string_or_die(&sig_filename, "%s.signed", data_filename);
+	/* create the cert store and set the verify callback */
+	if (!(store = X509_STORE_new())) {
+		goto error;
+	}
 
-	unlink(sig_filename);
+	X509_STORE_set_verify_cb_func(store, verify_callback);
 
-	free(sig_filename);
-}
+	/* Add the certificates to be verified to the store */
```

Comment:
The store created here is being used in the validate_signature() call as well.  I'd move store creation to it's own function, especially considering we may remove the call to validate_certificate till some future date